### PR TITLE
New Rule: ObjectLiteralToLambda

### PR DESCRIPTION
**Pull request by zmunm**
apply #3528 

I hope there isn't much I've missed.

Thank you @cortinico

### CODE REVIEW

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ *     }
+ * }
+ * </noncompliant>
+ *
+ * <compliant>
+ * Foo {
+ * }
+ * </compliant>
+ *
+ * @active since v1.17.0
```
Comment (BraisGabin) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:37]: Don't activate it by default please.

Comment (zmunm) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:37]: Thank you. Are there any criteria to set the rule to be enabled by default?

Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:37]: We never enable a rule by default the first time we introduced it because it can have false positives and we will impact to a lot of users. We prefer to enable it later if we think that its a rule that nearly all our user base will like to have it. But we don't have any defined rule here.

**Review by BraisGabin — APPROVED** [review]

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ *     }
+ * }
+ * </noncompliant>
+ *
+ * <compliant>
+ * Foo {
+ * }
+ * </compliant>
+ */
+class ObjectLiteralToLambda(config: Config = Config.empty) : Rule(config) {
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:37]: You should add the `@RequiresTypeResolution` annotation here.

**Review by cortinico — COMMENTED** [review]
Nice test coverage 👍

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    override val issue = Issue(
+        javaClass.simpleName,
+        Severity.Style,
+        "Report object literals that can be changed to lambdas.",
+        Debt.FIVE_MINS
+    )
+
+    private val KotlinType.isSamInterface
```
Comment (chao2zhang) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:47]:
```suggestion
    private val KotlinType.couldBeSamInterface
```

According to its [original documentation](
https://github.com/JetBrains/kotlin/blob/master/core/descriptors/src/org/jetbrains/kotlin/descriptors/ClassDescriptor.java#L112), `isDefinitelyNotSamInterface` may return falsely false

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import org.jetbrains.kotlin.psi.KtObjectLiteralExpression
+import org.jetbrains.kotlin.psi.KtSuperTypeEntry
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.types.KotlinType
+
+/**
+ * An anonymous object that does nothing other than the implementation of a single method
+ * can be used as a lambda.
+ *
+ * See https://kotlinlang.org/docs/fun-interfaces.html
```
Comment (chao2zhang) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:23]: Would this rule detect findings when the consuming code is Kotlin 1.3 or 1.3-?

Comment (chao2zhang) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:23]: (It may be good at least to call it out in the documentation)

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+
+class ObjectLiteralToLambdaSpec : Spek({
+    setupKotlinEnvironment()
+
+    val env: KotlinCoreEnvironment by memoized()
+    val subject by memoized { ObjectLiteralToLambda() }
+
+    describe("ObjectLiteralToLambda rule") {
+
+        context("report convertible expression") {
```
Comment (chao2zhang) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:20]: Would you mind adding a test case like the following? Note: This is from real production code that I found is not possible to use SAM Interface. This can be a follow-up PR since this may require more research or changes.
```
fun <T> LiveData<Resource<T>>.observeUntilFinished(observer: Observer<Resource<T>>? = null) {
    // Can't use SAM we are using implicit `this` receiver
    observeForever(object : Observer<Resource<T>> {
        override fun onChanged(resource: Resource<T>?) {
            observer?.onChanged(resource)
            if (resource?.status == Status.SUCCESS || resource?.status == Status.ERROR) {
                removeObserver(this)
            }
        }
    })
}
```

**Comment by chao2zhang** [comment]
Actually, I also find that converting to SAM may indicate anonymous object converted to singleton:  https://github.com/JetBrains/kotlin/commit/99a6bdeec7f052b180d859f59cbc5829ab6e9034

I guess this is an edge case but we should at least call it out. (We can soon have @Controverial annotation)

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt (hunk 5 above)
Comment (zmunm) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:20]: I've found some edge cases related to this, more research is needed, and code structure may change as well.
As of now, I cannot confirm only positive cases, so I would like to do more research.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt (hunk 2 above)
Comment (marschwar) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:39]: It is not merged yet.

Comment (cortinico) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:39]: Yup sorry, so let's sync and make sure that @marschwar we merge your PR first

Comment (chao2zhang) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:39]: `@RequiresTypeResolution` is now merged https://github.com/detekt/detekt/pull/3579

Comment (zmunm) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:46]: done. I was mistaken for KDoc  comment :sweat_smile:

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt (hunk 5 above)
Comment (zmunm) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:20]: done

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt (hunk 3 above)
Comment (zmunm) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt]: done

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt (hunk 4 above)
Comment (zmunm) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:31]: I wandering looking for a way to test. Can you give me a hint? @chao2zhang

**Comment by zmunm** [comment]
In the process of adding the case, there have been many modifications to the code and JetBrains/kotlin@99a6bde is still under research.
So I changed PR to a draft and I'll request again as soon as it is resolved. Thank you

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+    
+                fun newObject() = object : Sam {
+                    override fun foo() {
+                    }
+                }
+    
+                fun lambda() = Sam {}
+
+                val a = newObject() === newObject() // false
+                val b = lambda() === lambda() // true
```
Comment (zmunm) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:500]: I made a reproduction for JetBrains/kotlin@99a6bde

And I've found a bit past [issue](https://stackoverflow.com/questions/47025233/android-lifecycle-library-cannot-add-the-same-observer-with-different-lifecycle) that's occurring for similar reasons.

This is obviously confusing, but I haven't found a good solution to check this @chao2zhang

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt (hunk 4 above)
Comment (chao2zhang) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:31]: I believe this requires us to set up a sample project and set its kotlin version to 1.3.*

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt (hunk 5 above)
Comment (chao2zhang) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:20]: 👍  Just to confirm, did you refer to the code https://github.com/JetBrains/kotlin/blob/29b23e79f32791e456a5b4a453277f0f0b3e984d/idea/src/org/jetbrains/kotlin/idea/intentions/ObjectLiteralToLambdaIntention.kt#L83 to exclude this-reference?

Comment (zmunm) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:20]: Yes. I used that here.
https://github.com/detekt/detekt/pull/3599/files#diff-40f82782290e1472fb9c77f556f94156050d6f123b9eaa2e4da8edd62155854aR65

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt (hunk 4 above)
Comment (zmunm) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:31]: When running detektMain with java.util.Runnable in 1.3.72, it was confirmed that it is detected as a test case.

Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:31]: Couldn't we get that information from `CompilerResources`?

About how to test that `if`... we don't have any way to make it right now. But maybe we can create an issue to fix that and accept that `if` untested for now.

**Review by BraisGabin — APPROVED** [review]

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt (hunk 4 above)
Comment (zmunm) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:31]: Everything is new, so I'm grasping it. :loading: 
Please let me know if you have any misunderstandings.

From what I checked, in `CompileResource`, We can only see that the version is 1.3, and it seems that We can't actually compile with 1.3. Is that right?

What we want to check is the latter, and in java-interop there is nothing we can do with `LanguageVersion.KOTLIN_1_3` because the code we need to detect is exactly the same.

When I tested `fun interface` in 1.3, there is nothing else to do because it fails in the :compileKotlin task.

If I make an `issue`, I think I will have to rewrite the case in the module implementing 1.3, but I wonder if this falls within the scope of the project's support.

Comment (chao2zhang) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:31]: The bottom line is that we do not crash when compiling Kotlin 1.3.
What's better is that even if the code does not compile, detekt findings should be easy to understand.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+                val code = """
+                fun interface Sam {
+                    fun foo()
+                }
+                val a = object : Sam {
+                    override fun foo() {
+                    }
+                }
+                """
+                subject.compileAndLint(code).assert().isEmpty()
```
Comment (chao2zhang) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:156]:
```suggestion
                subject.compileAndLintWithContext(env, code).assert().isEmpty()
```
We should use compileAndLintWithContext() instead of compileAndLint() here, because the latter runs without type resolution and would degrade to no-op.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+                    }
+                }
+                """
+                subject.compileAndLintWithContext(env, code)
+                    .assert()
+                    .hasSize(1)
+                    .hasSourceLocations(SourceLocation(7, 9))
+            }
+
+            it("recursive call") {
```
Comment (chao2zhang) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:462]: Nice!

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt (hunk 6 above)
Comment (chao2zhang) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:500]: No worries. The test code is good documentation itself.

**Review by chao2zhang — APPROVED** [review]

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+                val code = """
+                fun interface Sam {
+                    fun foo()
+                }
+                val a = object : Sam {
+                    override fun foo() {
+                    }
+                }
+                """
+                subject.compileAndLintWithContext(env, code).assert().isEmpty()
```
Comment (chao2zhang) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:156]:
```suggestion
                subject.compileAndLint(code).assert().isEmpty()
```

My bad, I didn't realize that this test is for non-typeresolution.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+                """
+                subject.compileAndLintWithContext(env, code)
+                    .assert()
+                    .hasSize(1)
+                    .hasSourceLocations(SourceLocation(4, 9))
+            }
+        }
+
+        context("is not correct implement") {
+            it("without context") {
```
Comment (chao2zhang) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambdaSpec.kt:146]:
```suggestion
            it("without type resolution") {
```

**Review by chao2zhang — APPROVED** [review]

**Comment by zmunm** [comment]
@chao2zhang Please change the linked issue to #3528

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt (hunk 4 above)
Comment (chao2zhang) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/ObjectLiteralToLambda.kt:31]: Currently, the default configured `detekt` task does not depend on `compileKotlin`. If the error message is confusing, we should consider adding a `LanguageVersion.KOTLIN_1_3` version check, otherwise, this PR should be good to go.

**Comment by chao2zhang** [comment]
Should Detekt support primarily the current and the immediate previous feature release? 
As Kotlin 1.5 is approaching, adding checks for Kotlin 1.4 language features should be okay to merge and we can loose our compatibility requirement for Kotlin 1.3-.

**Comment by zmunm** [comment]
Is there something else I need to add for this?

**Comment by chao2zhang** [comment]
Kotlin 1.5 has arrived already (Not in Detekt yet), we could say that Detekt at maximum will be supporting 2 major releases of Kotlin.
